### PR TITLE
rfqmsg: request fields blip align

### PR DESCRIPTION
**Pull request by ffranr**
In this PR, we align the request RFQ wire message fields with the tap BLIP.

All TLV record values are already aligned with the BLIP due to previous PRs. This PR introduces internal changes to `tapd` to ensure the `request` message fields are handled, named, and interpreted correctly. These changes are non-interface and non-breaking.

This will be the last RFQ BLIP alignment PR.

### CODE REVIEW

**Comment by Roasbeef** [comment]
We should also make sure the min+max fields are set properly, along with the transfer type.

**Comment by guggero** [comment]
We should probably also remove the `--experimental.rfq.mockoraclesatsperasset=` flag in this PR (or a related PR).

**Comment by ffranr** [comment]
> We should probably also remove the `--experimental.rfq.mockoraclesatsperasset=` flag in this PR (or a related PR).

@guggero I would prefer to do that in a future PR. I want to prioritize BLIP alignment and address any bugs arising from misalignment issues.

File: rfq/order.go
Comment (Roasbeef) [inline rfq/order.go:272]: I think the constructor can still be used here?

File: rfqmsg/request.go
Comment (guggero) [inline rfqmsg/request.go]: :pray: thanks for re-ordering.

File: asset/asset.go
Comment (guggero) [inline asset/asset.go:287]: We have `fn.MaybeSome()` (which we should port over to `lnd` as well since it's quite useful).

File: asset/asset.go
Comment (guggero) [inline asset/asset.go:364]: Since we now kind of have a semantic distinction between the asset being specified (`fn.Option`) or all zero (for specifying an asset is BTC), shouldn't we also make the distinction here?
Meaning that if the asset ID isn't set, we'd print `id=, group_pub_key`?

File: rfq/negotiator.go
Comment (guggero) [inline rfq/negotiator.go:409]: Should we instead pass in the `Specifier` into `queryBidFromPriceOracle` to avoid some duplication? Then we likely also don't have to unwrap to pointer?

File: rfqmath/convert_test.go
Comment (guggero) [inline rfqmath/convert_test.go:501]: nit: can use `math.Pow10`.

File: rfqmsg/records_test.go
Comment (guggero) [inline rfqmsg/records_test.go:56]: nit: looks like this would fit on previous line?

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:409]: `queryBidFromPriceOracle` is called given a `BuyRequest` in another part of the code. And the `BuyRequest` does not have an `AssetSpecifier` field yet. I will add and then `queryBidFromPriceOracle` can accept an `asset.Specifier`.

File: rfq/order.go
Comment (ffranr) [reply rfq/order.go:272]: In the latest commits, I've introduced a new fixed-point method, `SetIntValue`, which enables constructing a fixed-point instance from any integer value of the same type as the fixed-point coefficient's type:
```
assetAmtFp := new(rfqmath.BigIntFixedPoint).SetIntValue(assetAmt)
```

**Comment by ffranr** [comment]
I think we should drop the `TransferType` field from the request message. As I was adding it I found that it was pulling in unnecessary context which I found made things more complicated. It relates different request messages to what there intended big picture uses might be. Which is not something the request/accept/reject RFQ messages need to be aware of IMO. It's a bit like a "note" or "label" but doesn't seem to add much when applied to the message level.

What do you think @guggero ?

**Comment by ffranr** [comment]
Things left to do in this PR:

I've added the `MinInAsset` and `MinOutAsset` request message fields. I think we need to add reject message error types for these fields as the accept message doesn't propose other lower bounds.

And we need to reject based on the dust limit in relation to `MinInAsset` or `MinOutAsset` depending on whichever one is BTC. Not sure where the dust limit is defined so might add my own msats dust limit const.

Any other use of these fields will probably be rewritten by the MPP OKR.

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:410]: I notice now that `BuyOrder` will need the asset specifier upgrade at the same time as `queryBidFromPriceOracle`. I'm just going to leave that for a future PR. It distracts from the main goal of this PR, IMO.

**Comment by ffranr** [comment]
I think we should merge this PR as is, subject to reviews.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:95]: IMO this name is a bit confusing. Since in the RFQ doc and on the wire we actually talk about the `max_in_asset`... 

In my mind it would be more consistent if this was called `MaxInAsset` here, the asset buy policy as well as in the `BuyRequest` and `SellRequest` messages.

OR do the switch at the transition from wire to request struct, but then have the units correct on the request structs (one being `uint64` and the other being `lnwire.MilliSatoshi`).
And the policies should 100% match the request structs (both in name and unit type), otherwise it'll be even more confusing.

**Review by guggero — COMMENTED** [review]
Thanks for all the updates. I think we are fairly close here, but I have a couple of things that we should try to get in with this PR, as IMO the changes are within the scope of the PR:

1. Rename the `SellRequest`'s `AssetAmount` either to `MaxInAsset` with `uint64` or to `MaxInboundAmount` with `lnwire.MilliSatoshi` (depending on the decision where to make the switch, see inline comment).
2. Enforce the above max amount in the asset purchase policy.
3. Since we don't have the "straightening out" of the buy/ask/purchase terminology I had in my WIP branch, and we still turn a buy request into an ask request for the oracle (which internally is then called a sale transaction), we should add that to the RFQ doc, best with a sequence diagram. Otherwise it will endlessly confuse me, and probably any authors of oracle logic as well.

**Comment by guggero** [comment]
> I think we should drop the `TransferType` field from the request message.

We currently decide whether it's a buy or sell request based on which asset is BTC here:
https://github.com/lightninglabs/taproot-assets/blob/e1e8268049f1dcc46b29df48c699ad5891bfa3c4/rfqmsg/request.go#L503

Once we support asset-to-asset conversion directly, how would we distinguish between buy and sell? Just the _direction_ of the in vs. out asset?

**Comment by ffranr** [comment]
> > I think we should drop the `TransferType` field from the request message.
> 
> We currently decide whether it's a buy or sell request based on which asset is BTC here:
> 
> https://github.com/lightninglabs/taproot-assets/blob/e1e8268049f1dcc46b29df48c699ad5891bfa3c4/rfqmsg/request.go#L503
> 
> Once we support asset-to-asset conversion directly, how would we distinguish between buy and sell? Just the _direction_ of the in vs. out asset?

IMO we should also drop the buy/sell types when we move to support asset-to-asset. So the transfer type is not needed then either.

File: rfq/order.go
Comment (ffranr) [reply rfq/order.go:95]: `order.go` is completely from the point of view of an edge node. That's one reason why we're using "policy" structs here and copying over fields rather than just carrying along an entire accepted quote.

I think things would be more confusing if we write everything from a wallet node's PoV including the edge node only routines.

File: taprpc/priceoraclerpc/price_oracle.proto
Comment (guggero) [inline taprpc/priceoraclerpc/price_oracle.proto:127]: Do we need this field for an actual use case today? I think we should just remove it to make things easier to understand... See my other comments as well.

File: rfq/negotiator.go
Comment (guggero) [inline rfq/negotiator.go:181]: This TODO remains unresolved and we don't pass any amount to the oracle.

File: rpcserver.go
Comment (guggero) [inline rpcserver.go:6956]:
```
		// paymentMaxAmt is the maximum amount that the counterparty is
		// expected to pay. This is the amount that the invoice is
		// asking for plus the fee limit in milli-satoshis.
		var paymentMaxAmt lnwire.MilliSatoshi
		if invoice.MilliSat == nil {
			amt, err := lnrpc.UnmarshallAmt(pReq.Amt, pReq.AmtMsat)
			if err != nil {
				return fmt.Errorf("error unmarshalling " +
					"amount: %w", err)
			}
			if amt == 0 {
				return errors.New("amount must be " +
					"specified when paying a zero amount " +
					"invoice")
			}

			paymentMaxAmt = amt
		} else {
			paymentMaxAmt = *invoice.MilliSat
		}

		// Calculate the fee limit that should be used for this payment.
		feeLimit, err := lnrpc.UnmarshallAmt(
			pReq.FeeLimitSat, pReq.FeeLimitMsat,
		)
		if err != nil {
			return fmt.Errorf("error unmarshalling fee limit: %w",
				err)
		}

		paymentMaxAmt += feeLimit
```

**Review by guggero — COMMENTED** [review]
Almost there, thanks a lot for sticking to this huge undertaking :100:

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:168]: This TODO is trying to say that `buyOrder` should have an `asset.Specifier` type field rather than constructing an asset specifier at this point. I will clarify the todo in code.

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:109]: > Is the assetMaxAmount used anywhere anymore after this series of commits?

`assetMaxAmt` is passed on to `n.cfg.PriceOracle.QueryBidPrice` below. It gets passed to the price oracle service.

> Can't we just remove it, and make the `paymentMaxAmount` a non-option and just use that everywhere?

These two fields relate to two different assets.

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:181]: Adding and populating that field in the `BuyOrder` would require an RPC change. I mean to do that in a future PR. I think this PR is getting too long.

File: taprpc/priceoraclerpc/price_oracle.proto
Comment (ffranr) [reply taprpc/priceoraclerpc/price_oracle.proto:127]: `MaxInAsset` sometimes relates to BTC and sometimes to a tap asset.

The two different fields in the price oracle relate to the two different assets that we're tying to find an exchange rate for. Instead of a single field that is sometimes the upper bound on the tap asset and sometimes an upper bound on the BTC side, I've got two different fields.

File: rpcserver.go
Comment (ffranr) [reply rpcserver.go:6956]: Thank you!

File: taprpc/priceoraclerpc/price_oracle.proto
Comment (guggero) [reply taprpc/priceoraclerpc/price_oracle.proto:127]: Okay, I guess I see where my confusion is coming from. It's because we still do the "flip" of the transaction type for the oracle (which I eliminated in my WIP refactor PR).

What I mean is:
 - On the wallet end user side, a `SellOrder` goes into the oracle as an `ask` (transaction type `sale` with `payment_assset_max_amount` set)
 - On the edge node, that same `SellOrder` goes into the oracle as a `bid` (transaction type `purchase` with `payment_max_amount` set)

And then for a `BuyOrder` it would be kind of the other way around.
Can we please add all of this to the RFQ documentation?

And perhaps update the comments in this proto file to make it more clear to the developer of an oracle when what field is set? For example that never both `_max_amount` fields are set at the same time?

File: rfq/negotiator.go
Comment (guggero) [reply rfq/negotiator.go:109]: Okay, I see. Confusion on my part, see comment below with a request for documenting this a bit more.

File: rfq/negotiator.go
Comment (guggero) [reply rfq/negotiator.go:182]: But at this specific call site, we only pass in `fn.None` for all the values. So we're not passing any amount when requesting a rate for the hint. Is that an allowed way to call into the oracle, with none of the `_max_amount` fields set?

**Review by guggero — APPROVED** [review]
Okay, I think I finally understand all the changes and aligned my own mental model to them. Would love to have a bit more documentation around that, but not blocking this PR on it. So LGTM, nice work! :tada:

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:182]: I think it should technically be a valid way of calling the price oracle. If upper bounds aren't given then the price would just be less accurate. But we should complete the TODO. I'll add the correct field to `BuyOrder` in a separate PR.

**Comment by ffranr** [comment]
Further work to be included in a separate PR:

* Better documentation for `SellRequest`, `BuyRequest`, `BuyOrder`, `SellOrder`.
* Add missing field to `BuyOrder` and populate in RPC endpoint.

**Comment by guggero** [comment]
> Further work to be included in a separate PR:
> 
>     * Better documentation for `SellRequest`, `BuyRequest`, `BuyOrder`, `SellOrder`.
> 
>     * Add missing field to `BuyOrder` and populate in RPC endpoint.


Sure, let's do all of that in a follow-up PR, then we can discuss there.

> @guggero are you also ok with the following:
> 
>     * Remove transfer type field from `request` RFQ message in BLIP and revise TLV integers in BLIP and tapd.

I still don't fully see how that would work, given that we'd still need to know whether to ask the oracle for an ask or bid. Or would we also remove that directionality there and only have a single RPC method?

**Comment by ffranr** [comment]
> I still don't fully see how that would work, given that we'd still need to know whether to ask the oracle for an ask or bid. Or would we also remove that directionality there and only have a single RPC method?

Yes, I think we only need a single oracle interface method. The oracle needs to know what asset is inbound and which is outbound and then it can return the ask/bid based on that. We can do that now with the oracle RPC endpoint by setting payment asset, subject asset, and the transfer type. But we can simplify the RPC endpoint's messages further also.

File: taprpc/priceoraclerpc/price_oracle.proto
Comment (GeorgeTsagk) [inline taprpc/priceoraclerpc/price_oracle.proto:131]: are we ok with breaking this?

File: rpcserver.go
Comment (GeorgeTsagk) [inline rpcserver.go:6603]: shouldn't this TODO be addressed before merging?

**Review by GeorgeTsagk — APPROVED** [review]
Very clean commits :100: 

LGTM, straight forward changes :white_check_mark: 

have some minor Qs/nits, non-blocking

File: rpcserver.go
Comment (ffranr) [reply rpcserver.go:6603]: I'll put up another PR to address this. This PR is getting a bit long.

**Pull request merged**
